[Security Solution] Adds `customized_fields` and `has_base_version` fields to `rule_source` object schema

dplumlee opened the pull request:
**Resolves: https://github.com/elastic/security-team/issues/12507** (internal)

## Summary

Adds two new fields to the existing `rule_source` object in our rule schema as described in https://github.com/elastic/kibana/pull/230856. Also updates and adds test coverage for the new field logic.

The new fields are:

- `customized_fields`: an array of objects containing rule field names that have been modified from the base version of the prebuilt rule.
  - Defaults to empty array if prebuilt rule is not customized or if base version did not exist during diff calculation.
- `has_base_version`: a boolean field that specifies if the base version of a prebuilt rule was able to be fetched and used during the customization calculation.

This PR also adds related telemetry fields as described in https://github.com/elastic/kibana/pull/230856. This includes a `customizations` object field which contains a slimmed down version of `customized_fields` and has a `num_functional_fields` number field that is created in the telemetry task pipeline by comparing the customized fields array to a constant list of field names that we are defining as "functional". This source of truth list can be found in the `x-pack/solutions/security/plugins/security_solution/common/detection_engine/constants.ts` file

### Examples

```json
{
  "rule_source": {
    "type": "external",
    "is_customized": true,
    /* New fields */
    "customized_fields": [
      {
        "field_name": "tags",
      },
      {
        "field_name": "query",
      }
    ],
    "has_base_version": true
  }
}
```

```json
"customizations": {
  "customized_fields": ["tags", "query"],
  "num_functional_fields": 2,
}
```

## How to test telemetry

Link to internal staging with example data: ([internal staging](https://analytics-staging.sde.elastic.dev/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:60000),time:(from:'2025-09-26T15:59:24.512Z',to:'2025-09-26T16:08:58.435Z'))&_a=(columns:!(),dataSource:(dataViewId:'4ca97040-d095-11ec-95a5-011050c1180f',type:dataView),filters:!(),interval:auto,query:(language:kuery,query:'customizations.num_functional_fields%20%3E%200'),sort:!(!('@timestamp',desc)),viewMode:documents)))

1. Set the prebuilt rule task type to something shorter than `1hr` in this file: `x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/tasks/prebuilt_rule_alerts.ts`
2. Add the following to `kibana.dev.yml`:
```
telemetry.enabled: true
telemetry.optIn: true

// (Optional for checking to see if its working)
logging:
  root:
    appenders: [default]
    level: warn
  loggers:
    - name: plugins.securitySolution
      level: debug
    - name: plugins.ruleRegistry
    - name: plugins.taskManager
```
3. Start up both Elasticsearch and kibana (Has to be done _after_ updating task interval as task objects are stored in ES)
4. Install prebuilt rules
5. Modify prebuilt rules with different field customizations and enable those rules
6. Generate alerts that match these rules (resolver script generator, dev tools, query modification, etc.)
7. View the alerts getting sent to the internal staging telemetry cluster (https://analytics-staging.sde.elastic.dev) in the `detections_alert_telemetry_elastic*` index
8. Use the new `customizations` field to filter out/in customized rule alerts

## Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
  - [x] [Rule customization tests](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/9317)


<!--ONMERGE {"backportTargets":["9.1","9.2"]} ONMERGE-->

dplumlee committed:
adds field

fixes tests

runs make

updates tests

cleans up types

fix tests

updates tests

updates snapshots

fixes tests

fixes tests

fixes tests

fixes tests

addresses comments

language updates

adds logic guard for missing fields

updates new tests

addresses comments

[CI] Auto-commit changed files from 'make api-docs'

address non-test comments

fixes tests

jkelas reviewed (commented):
Hi @dplumlee 
I reviewed your changes and the code looks very good. Well done.
I only left two minor comments.

I did the manual testing and all works fine with the modified functionality. I also observed the changes made in the `kibana_alerting_cases` index for modified rules and everything worked as expected.

jkelas reviewed (approved):
Thank you @dplumlee for adressing my comments.
I am approving.

natasha-moore-elastic commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml:
```diff
@@ -70,6 +70,23 @@ components:
 
+    ExternalRuleHasBaseVersion:
+      type: boolean
+      description: Determines whether or not an external/prebuilt rule has its original, unmodified version present when the `is_customized` calculation is performed.
```
```suggestion
      description: Determines whether an external/prebuilt rule has its original, unmodified version present when the `is_customized` calculation is performed.
```
Not sure what 'external' rule means – AFAIK, that's not a term we use in our user docs.

dplumlee replied:
It is another name for prebuilt rules, we had already used the naming in the [current docs](https://www.elastic.co/docs/api/doc/kibana/operation/operation-findrules#operation-findrules-200-body-application-json-data-rule_source) for the rule source object so I just kept it the same.

natasha-moore-elastic commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml:
```diff
@@ -70,6 +70,23 @@ components:
+        properties:
+          field_name:
+            type: string
+            description: Name of the field in the rule object that has been customized from its base value by the user.
```
```suggestion
            description: Name of a user-modified field in the rule object.
```

natasha-moore-elastic commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/rule_schema/common_attributes.schema.yaml:
```diff
@@ -70,6 +70,23 @@ components:
+        required:
+          - field_name
+
+      description: An array of field names that have been customized by the user (i.e. any of its fields that have been modified and have diverged from the base value). Defaults to an empty array.
```
```suggestion
      description: An array of customized field names—that is, fields that the user has modified from their base value. Defaults to an empty array.
```

natasha-moore-elastic reviewed (approved):
Left some optional wording suggestions, otherwise LGTM!

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/339487)
* Commit: c73a99cc2fcdffd2743f45457471f50767e9d6b0

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b3a75dc0b29dd9d9db4658e5235a86d4c35a6f28) | [after](https://github.com/elastic/kibana/commit/c73a99cc2fcdffd2743f45457471f50767e9d6b0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 10.5MB | 10.5MB | +110.0B |



### History
* :green_heart: [Build #339118](https://buildkite.com/elastic/kibana-pull-request/builds/339118) succeeded 09f251fa9e4aa45615ef74132b68355d439319eb
* :green_heart: [Build #338914](https://buildkite.com/elastic/kibana-pull-request/builds/338914) succeeded b0dda458b32cf17709b0859fd264f83e085cdf41
* :broken_heart: [Build #338590](https://buildkite.com/elastic/kibana-pull-request/builds/338590) failed 79dccbb5873c2c726327269b517f98b7e6adb4d7
* :broken_heart: [Build #338587](https://buildkite.com/elastic/kibana-pull-request/builds/338587) failed e5b7e10b9bfb539f394358d72b13e2abc41eb599
* :broken_heart: [Build #338535](https://buildkite.com/elastic/kibana-pull-request/builds/338535) failed 8f1c63cb4d39867a8ef99eeda749b3cc2e5840f2
* :broken_heart: [Build #338510](https://buildkite.com/elastic/kibana-pull-request/builds/338510) failed 058b19df385c5383de5f72991aae64ac817ccb3f

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339487","number":339487,"commit":"c73a99cc2fcdffd2743f45457471f50767e9d6b0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339118","number":339118,"commit":"09f251fa9e4aa45615ef74132b68355d439319eb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338914","number":338914,"commit":"b0dda458b32cf17709b0859fd264f83e085cdf41"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338590","number":338590,"commit":"79dccbb5873c2c726327269b517f98b7e6adb4d7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338587","number":338587,"commit":"e5b7e10b9bfb539f394358d72b13e2abc41eb599"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338535","number":338535,"commit":"8f1c63cb4d39867a8ef99eeda749b3cc2e5840f2"}],"number":339487}
buildkite-pr-comment-->

denar50 reviewed (approved):
Code review only. LGTM for changes to the code owned by DE.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/339840)
* Commit: 39e0ef7362b71504e99082f0fdbc6c897f45a0d5

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/339840#01995296-703d-4bcc-9f77-7d62522c8927)

### History
* :green_heart: [Build #339487](https://buildkite.com/elastic/kibana-pull-request/builds/339487) succeeded c73a99cc2fcdffd2743f45457471f50767e9d6b0
* :green_heart: [Build #339118](https://buildkite.com/elastic/kibana-pull-request/builds/339118) succeeded 09f251fa9e4aa45615ef74132b68355d439319eb
* :green_heart: [Build #338914](https://buildkite.com/elastic/kibana-pull-request/builds/338914) succeeded b0dda458b32cf17709b0859fd264f83e085cdf41
* :broken_heart: [Build #338590](https://buildkite.com/elastic/kibana-pull-request/builds/338590) failed 79dccbb5873c2c726327269b517f98b7e6adb4d7
* :broken_heart: [Build #338587](https://buildkite.com/elastic/kibana-pull-request/builds/338587) failed e5b7e10b9bfb539f394358d72b13e2abc41eb599
* :broken_heart: [Build #338535](https://buildkite.com/elastic/kibana-pull-request/builds/338535) failed 8f1c63cb4d39867a8ef99eeda749b3cc2e5840f2

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339840","number":339840,"commit":"39e0ef7362b71504e99082f0fdbc6c897f45a0d5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339487","number":339487,"commit":"c73a99cc2fcdffd2743f45457471f50767e9d6b0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339118","number":339118,"commit":"09f251fa9e4aa45615ef74132b68355d439319eb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338914","number":338914,"commit":"b0dda458b32cf17709b0859fd264f83e085cdf41"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338590","number":338590,"commit":"79dccbb5873c2c726327269b517f98b7e6adb4d7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/338587","number":338587,"commit":"e5b7e10b9bfb539f394358d72b13e2abc41eb599"}],"number":339840}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340037)
* Commit: 9414a0f838aab75ab478db746e61cd186d0634c9
* [Cloud Deployment](https://kibana-pr-234793-dfa1b3.kb.us-west2.gcp.elastic-cloud.com)

### History
* :broken_heart: [Build #340006](https://buildkite.com/elastic/kibana-pull-request/builds/340006) failed a14f8cad49d50821bf89ce5c18db5f283330a418
* :broken_heart: [Build #339895](https://buildkite.com/elastic/kibana-pull-request/builds/339895) failed b23ff7fe2c14744df671c77f5bbc6f23900b45a8
* :broken_heart: [Build #339840](https://buildkite.com/elastic/kibana-pull-request/builds/339840) failed 39e0ef7362b71504e99082f0fdbc6c897f45a0d5
* :green_heart: [Build #339487](https://buildkite.com/elastic/kibana-pull-request/builds/339487) succeeded c73a99cc2fcdffd2743f45457471f50767e9d6b0
* :green_heart: [Build #339118](https://buildkite.com/elastic/kibana-pull-request/builds/339118) succeeded 09f251fa9e4aa45615ef74132b68355d439319eb
* :green_heart: [Build #338914](https://buildkite.com/elastic/kibana-pull-request/builds/338914) succeeded b0dda458b32cf17709b0859fd264f83e085cdf41

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340037","number":340037,"commit":"9414a0f838aab75ab478db746e61cd186d0634c9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340006","number":340006,"commit":"a14f8cad49d50821bf89ce5c18db5f283330a418"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339895","number":339895,"commit":"b23ff7fe2c14744df671c77f5bbc6f23900b45a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339840","number":339840,"commit":"39e0ef7362b71504e99082f0fdbc6c897f45a0d5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339487","number":339487,"commit":"c73a99cc2fcdffd2743f45457471f50767e9d6b0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339118","number":339118,"commit":"09f251fa9e4aa45615ef74132b68355d439319eb"}],"number":340037}
buildkite-pr-comment-->

banderror reviewed (commented):
@dplumlee Sorry, I wasn't able to finish my review today. Posting the comments I already have and will continue tomorrow morning.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340130)
* Commit: f2812ce0e3aa5b71e39c84f31651aca6f42a96e0
* [Cloud Deployment](https://kibana-pr-234793-dfa1b3.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #340037](https://buildkite.com/elastic/kibana-pull-request/builds/340037) succeeded 9414a0f838aab75ab478db746e61cd186d0634c9
* :broken_heart: [Build #340006](https://buildkite.com/elastic/kibana-pull-request/builds/340006) failed a14f8cad49d50821bf89ce5c18db5f283330a418
* :broken_heart: [Build #339895](https://buildkite.com/elastic/kibana-pull-request/builds/339895) failed b23ff7fe2c14744df671c77f5bbc6f23900b45a8
* :broken_heart: [Build #339840](https://buildkite.com/elastic/kibana-pull-request/builds/339840) failed 39e0ef7362b71504e99082f0fdbc6c897f45a0d5
* :green_heart: [Build #339487](https://buildkite.com/elastic/kibana-pull-request/builds/339487) succeeded c73a99cc2fcdffd2743f45457471f50767e9d6b0

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340130","number":340130,"commit":"f2812ce0e3aa5b71e39c84f31651aca6f42a96e0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340037","number":340037,"commit":"9414a0f838aab75ab478db746e61cd186d0634c9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340006","number":340006,"commit":"a14f8cad49d50821bf89ce5c18db5f283330a418"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339895","number":339895,"commit":"b23ff7fe2c14744df671c77f5bbc6f23900b45a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339840","number":339840,"commit":"39e0ef7362b71504e99082f0fdbc6c897f45a0d5"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339487","number":339487,"commit":"c73a99cc2fcdffd2743f45457471f50767e9d6b0"}],"number":340130}
buildkite-pr-comment-->

ersin-erdal reviewed (changes requested):
Since this PR introduces a rule param change requires an intermediate release.

Can you create a PR that only adds the new params to the schema? That PR needs to be merged and deployed to serverless first.

dplumlee commented:
@ersin-erdal thanks for noticing this, I was not aware of the intermediate release protocol. I've added the fields as optional to the internal schema [here](https://github.com/elastic/kibana/pull/235394) and will wait until that gets merged/deployed to serverless before this goes in.

banderror commented:
@ersin-erdal Can you confirm that merging https://github.com/elastic/kibana/pull/235394 by the end of this week will unblock this PR on Monday next week?

I'm personally struggling to understand what issues are we trying to mitigate by doing this. Could you please explain what exactly we'd break if we were to merge this PR as is? Please note that the new fields introduced in this PR are not used anywhere yet.

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/mergers/rule_source/calculate_is_customized.ts:
```diff
@@ -43,22 +62,33 @@ export function calculateIsCustomized({
     // If the rule was previously customized, there's no way to determine
     // whether the customization remained or was reverted. Keeping it as
     // customized in this case.
-    return true;
+    return {
+      isCustomized: true,
+      customizedFields: [],
+    };
```
> how are we supposed to determine which fields are different from their base values? If name were reverted to its original value we'd have no way of determining that and removing it from the customizedFields array. We set it to an empty array as a default value

Yes, in this case, we wouldn't be able to determine if the name is still customized or has been reverted to its base version (which is not available anymore).

I agree now with marking it as `isCustomized: true` but `customizedFields: []` in this case. I think it'd be causing less confusion compared to keeping customized fields unchanged.

Thanks for the explanation @dplumlee 👍

banderror replied:
Still, my question is: can you point me to some test plan scenarios and integration tests that cover this edge case?

dplumlee replied:
I think most of the existing `x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/customization_enabled/customization/detect_customization_without_base_version.ts` test suite covers this but I also added a specific test to the suite for when the rule was previously customized and loses its base version. I can add that to the test plan scenario as well

banderror replied:
@dplumlee Perfect, this looks more than robust now 👍

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340853)
* Commit: 6cf2358602a91ceedff40f7d28d65246c074f5a5
* [Cloud Deployment](https://kibana-pr-234793-dfa1b3.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #340130](https://buildkite.com/elastic/kibana-pull-request/builds/340130) succeeded f2812ce0e3aa5b71e39c84f31651aca6f42a96e0
* :green_heart: [Build #340037](https://buildkite.com/elastic/kibana-pull-request/builds/340037) succeeded 9414a0f838aab75ab478db746e61cd186d0634c9
* :broken_heart: [Build #340006](https://buildkite.com/elastic/kibana-pull-request/builds/340006) failed a14f8cad49d50821bf89ce5c18db5f283330a418
* :broken_heart: [Build #339895](https://buildkite.com/elastic/kibana-pull-request/builds/339895) failed b23ff7fe2c14744df671c77f5bbc6f23900b45a8
* :broken_heart: [Build #339840](https://buildkite.com/elastic/kibana-pull-request/builds/339840) failed 39e0ef7362b71504e99082f0fdbc6c897f45a0d5

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340853","number":340853,"commit":"6cf2358602a91ceedff40f7d28d65246c074f5a5"},{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340130","number":340130,"commit":"f2812ce0e3aa5b71e39c84f31651aca6f42a96e0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340037","number":340037,"commit":"9414a0f838aab75ab478db746e61cd186d0634c9"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340006","number":340006,"commit":"a14f8cad49d50821bf89ce5c18db5f283330a418"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339895","number":339895,"commit":"b23ff7fe2c14744df671c77f5bbc6f23900b45a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339840","number":339840,"commit":"39e0ef7362b71504e99082f0fdbc6c897f45a0d5"}],"number":340853}
buildkite-pr-comment-->

banderror reviewed (commented):
@dplumlee Posting a few more comments, should wrap up the review soon.

banderror commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/import_with_installing_package.ts:
```diff
@@ -28,7 +28,6 @@ const NON_CUSTOMIZED_PREBUILT_RULE = PREBUILT_RULE_ASSET_A;
 const CUSTOMIZED_PREBUILT_RULE = {
   ...PREBUILT_RULE_ASSET_B,
   description: 'Custom description',
-  tags: ['custom-tag'],
```
Why was this removed?

banderror commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/customization_enabled/customization/customize_via_bulk_editing.ts:
```diff
@@ -340,7 +340,7 @@ export default ({ getService }: FtrProviderContext): void => {
 
     describe('when base version is available', () => {
-      testCustomizationViaBulkEditing({ hasBaseVersion: false });
+      testCustomizationViaBulkEditing({ hasBaseVersion: true });
```
👍

banderror reviewed (commented):
@dplumlee Posting the last chunk of comments which wraps my review. I'm not confident about:

1. All those back-and-forth normalization and denormalization functions. The code looks error-prone. Let's try to refactor.
2. Edge cases handling in the `calculateExternalRuleSource` function. Let's continue discussing this.
3. It would be good to have dedicated tests (primarily unit tests for the rule source calculation logic) for the newly added fields. I'm pretty sure we haven't covered some edge cases. Ideally we add these tests in this same PR. Refactoring the rest of the tests (primarily integration tests) to reduce the number of places where we had to add assertions for the new fields would be out of the scope.
4. When we're done with points 2 and 3, could you please collect a list of links to the tests, per each edge case / branch of the code of the `calculateExternalRuleSource` function. These should be dedicated tests for those cases, not some tests implicitly covering them.
5. I'd appreciate another pair of eyes on the correctness and completeness of the test coverage - @maximpn you should have the most expertise with these tests at this point.

pborgonovi commented:
@banderror I'll be full time on it and expect to be done by Monday.

pborgonovi commented:
Hey @banderror @dplumlee just sharing updates for today: I explored all workflows around rules customization, changes reversion, upgrade and import/export including rules with missing base version. 

I haven't come across any issue that could be related to this PR.

I've found 2 issues which I've reported:


[235838](https://github.com/elastic/kibana/issues/235838 ) (which @dplumlee has shared it's related to [214292](https://github.com/elastic/kibana/issues/214292)) and [235848](https://github.com/elastic/kibana/issues/235848) 

Both issues are currently present in production, so they're not side effects of this PR changes.

I'll give it another round Monday morning before wrapping up the tests.

ersin-erdal commented:
> @ersin-erdal Can you confirm that merging #235394 by the end of this week will unblock this PR on Monday next week?
> 
> I'm personally struggling to understand what issues are we trying to mitigate by doing this. Could you please explain what exactly we'd break if we were to merge this PR as is? Please note that the new fields introduced in this PR are not used anywhere yet.

Hi, 
It is documented here: https://docs.elastic.dev/reops/developer-guide/best-practices-rules#upgrades-and-rollbacks
Ying has added a GH action to inform the PR owners about the intermediate release requirement.

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

pborgonovi commented:
@banderror @dplumlee I haven't come across any issues other than the ones mentioned above. 
Please let me know if another round of exploratory tests is needed after further changes made to this PR.

dplumlee commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/mergers/rule_source/calculate_is_customized.ts:
```diff
@@ -43,22 +62,33 @@ export function calculateIsCustomized({
+  const customizedFields = getCustomizedFields(currentRule, nextRule);
+  return {
+    isCustomized: customizedFields.length > 0,
+    customizedFields: [], // Don't have base version so these fields are not necessarily the same fields that would be calculated from base_version
```
I'm not sure I'm following this proposal snippet, the way we have the logic branching right now in the PR is what I'd prefer to keep since it's effectively the same way rule source currently calculates, we're just adding the whole object as a return instead of just the `is_customized` boolean. The `internal` type branch in your comment should technically be unreachable since we already know the immutable field is true when we are at this point in the evaluation but with the current logic flow we would also catch that edge case and set it to `external`. And by defaulting to `is_customized: currentRule.rule_source.is_customized`, we would lose the ability to mark rules as customized when they lost their base version (unless previously customized with a base version) which would be a regression from what we have currently implemented.

As for the `customized_fields` field, I would propose to keep it only populating when the rule has a base version due to the reasons explained in my previous comment. I went ahead and added some more thorough comments describing this.

banderror replied:
@dplumlee Not sure I'm still following my proposal either 😆🤦‍♂️ I reviewed it again and the current logic makes sense to me.

This last block of code handles the case when the base version is not available, and the rule is not marked as customized (yet). If there's a change in the next rule revision, we mark the rule as customized, but don't persist the customized fields as we can't rely on their calculation due to the missing base version. If there's no change, we keep it as `is_customized: false`:

```ts
  const customizedFields = getCustomizedFields(currentRule, nextRule);
  return {
    type: 'external',
    is_customized: customizedFields.length > 0,
    customized_fields: [],
    has_base_version: false,
  };
```

Then, after the rule was customized, if the base version is still missing, any further updates to it will keep it as customized with the empty array of customized fields, because we can't know if these updates customize the rule further or reset it to its base version. And this is handled by the previous block of code:

```ts
  if (
    currentRule.rule_source.type === 'external' &&
    currentRule.rule_source.is_customized === true
  ) {
    // If the rule was previously customized, there's no way to determine
    // whether the customization remained or was reverted. Keeping it as
    // customized in this case.
    return {
      type: 'external',
      is_customized: true,
      customized_fields: [],
      has_base_version: false,
    };
  }
```

So, it seems we should keep it as is. Thank you for adding the comments.

dplumlee committed:
adds telemetry work

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

dplumlee committed:
lints

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/345309)
* Commit: d24715769ed532b386d5e900802d37998dd0e3f1

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/345309#01999694-9ffb-486a-9172-d2b26e735311)

### History
* :broken_heart: [Build #345302](https://buildkite.com/elastic/kibana-pull-request/builds/345302) failed 548af5374bac2abb4ad324ac78c2d646456d9122
* :broken_heart: [Build #345282](https://buildkite.com/elastic/kibana-pull-request/builds/345282) failed c9df76c7544ae659983fac51873b84d6917d7a0f
* :broken_heart: [Build #345259](https://buildkite.com/elastic/kibana-pull-request/builds/345259) failed d7a8aad2ed409a1967bf8b532bc6a8e63c9e0918
* :green_heart: [Build #342891](https://buildkite.com/elastic/kibana-pull-request/builds/342891) succeeded fd89059394a16da7f66d75257b7e077c43bca8df
* :broken_heart: [Build #342676](https://buildkite.com/elastic/kibana-pull-request/builds/342676) failed fd89059394a16da7f66d75257b7e077c43bca8df
* :broken_heart: [Build #342370](https://buildkite.com/elastic/kibana-pull-request/builds/342370) failed d6baa369161982c197ca290af2fb60f698712d53

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345309","number":345309,"commit":"d24715769ed532b386d5e900802d37998dd0e3f1"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345302","number":345302,"commit":"548af5374bac2abb4ad324ac78c2d646456d9122"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345282","number":345282,"commit":"c9df76c7544ae659983fac51873b84d6917d7a0f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345259","number":345259,"commit":"d7a8aad2ed409a1967bf8b532bc6a8e63c9e0918"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/342891","number":342891,"commit":"fd89059394a16da7f66d75257b7e077c43bca8df"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/342676","number":342676,"commit":"fd89059394a16da7f66d75257b7e077c43bca8df"}],"number":345309}
buildkite-pr-comment-->

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

dplumlee committed:
fixes tests on merge

dplumlee committed:
updates test plans

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

dplumlee commented:
@banderror 

> All those back-and-forth normalization and denormalization functions. The code looks error-prone. Let's try to refactor

I refactored quite a bit of this code so we separate the camel casing and normalization actions, I think it's a bit more clear. There needs to be further refactoring to strip out the remaining uses of the deprecated normalize functions but I thought that was outside of the scope of this PR as that function has been deprecated for a while now.

> When we're done with points 2 and 3, could you please collect a list of links to the tests, per each edge case / branch of the code of the calculateExternalRuleSource function. These should be dedicated tests for those cases, not some tests implicitly covering them.

I've added updates to the test plans themselves concerning all of these cases, as well as added unit tests that cover all logic paths of the `calculateExternalRuleSource` function. For API tests:

 - Import tests and edge cases can be found here: `x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export`
   - All suites have been updated to test for our new fields and the logic flow of an imported rule already existing or not.
 - Customizations w/ base version can be found here, testing for every field: `x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/customization_enabled/customization/detect_customization_with_base_version.ts`
 - Customizations w/o base version can be found here: `x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/customization_enabled/customization/detect_customization_without_base_version.ts`
   - I have added tests covering rules previously customized here

All other related tests and mocks have been updated to match the new logic and fields, we cover each of the 4 different paths in the `calculateExternalRuleSource` with both unit and FTR tests - many of these instances are just ensuring we reset the `customized_fields` value to an empty array and correctly label a base version as missing.

elasticmachine committed:
Merge branch 'main' into rule-customizations-field

ersin-erdal reviewed (approved):
LGTM
Since [the PR](https://github.com/elastic/kibana/pull/235394) for intermediate release has been merged and deployed, removing the blocker.

szaffarano reviewed (approved):
LGTM

JDKurma commented on x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/filterlists/prebuilt_rules_alerts.ts:
```diff
@@ -49,6 +49,9 @@ export const prebuiltRuleAllowlistFields: AllowlistFields = {
   'kibana.alert.rule.updated_at': true,
   'kibana.alert.rule.uuid': true,
   'kibana.alert.rule.version': true,
+  'kibana.alert.rule.parameters': {
```
Is `rule_source` nested under `kibana.alert.rule.parameters` or is it a flat key `kibana.alert.rule.parameters.rule_source`?

dplumlee replied:
It's nested under, `kibana.alert.rule.parameters` is the flat key for the whole parameters object but we didn't want to include everything there for document size reasons

dplumlee committed:
Merge branch 'main' into rule-customizations-field

banderror committed:
Fix rule normalization logic

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/converters/normalize_rule_params.ts:
```diff
@@ -4,44 +4,23 @@
-};
-
-export const normalizeRuleParams = (params: BaseRuleParams): NormalizedRuleParams => {
+export const normalizeRuleParams = (params: BaseRuleParams) => {
```
Fixed and pushed it in this commit: https://github.com/elastic/kibana/pull/234793/commits/14d09a9d878f4ff190629ec7d161981630b69fe6

dplumlee committed:
fixes tests

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/types.ts:
```diff
@@ -86,6 +87,9 @@ export interface TelemetryEvent {
       pod?: SearchTypes;
     };
   };
+  'kibana.alert.rule.parameters'?: {
+    rule_source?: RuleSource;
+  };
```
Let's specify the `customizations` field and its type in the `TelemetryEvent` as well.

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/helpers.ts:
```diff
@@ -389,6 +391,25 @@ export const processK8sUsernames = (clusterId: string, event: TelemetryEvent): T
   return event;
 };
 
+export const processDetectionRuleCustomizations = (event: TelemetryEvent) => {
+  const ruleSource = event['kibana.alert.rule.parameters']?.rule_source;
+  if (
+    !ruleSource ||
+    ruleSource.type === 'internal' ||
+    ruleSource.is_customized === false ||
+    ruleSource.has_base_version === false
+  ) {
+    return undefined; // Don't return anything if rule is not customized or base version doesn't exist
+  }
+  const numberOfFunctionalFields = ruleSource.customized_fields.filter(
+    (field) => FUNCTIONAL_FIELD_MAP[field.field_name as keyof RuleResponse]
+  ).length;
+  return {
+    customized_fields: ruleSource.customized_fields.map((fieldObj) => fieldObj.field_name),
+    num_functional_fields: numberOfFunctionalFields,
+  };
+};
```
This implementation doesn't handle the case when `customized_fields` is undefined. This will happen with any alerts generated prior to the release of this feature. It may even happen with alerts generated after releasing it, if they are generated by a rule which won't be updated by the user yet. When this happens, the function will throw an exception and break telemetry.

We should be very cautions about the data we process and code defensively.

Please add explicit null/undefined checks for both `customized_fields` and `has_base_version` fields and add corresponding test cases to `x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/helpers.test.ts`.

dplumlee replied:
Good catch

marshallmain replied:
This is one of the main reasons I implemented the versioned alerts schema: https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/alerts/schema.ts - it improves type safety in handling cases where we add new fields to alerts. Unfortunately, the rule schemas are not versioned so the alert schema doesn't get versioned fields for data copied from the rule. If we envision more scenarios like this in the future it could be worth investing in creating versioned rule schemas that can integrate into the alert schemas.

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/helpers.ts:
```diff
@@ -389,6 +391,25 @@ export const processK8sUsernames = (clusterId: string, event: TelemetryEvent): T
   return event;
 };
 
+export const processDetectionRuleCustomizations = (event: TelemetryEvent) => {
```
Please specify a return type explicitly

banderror reviewed (changes requested):
@dplumlee Blocking the PR as there are still some bugs to fix and feedback to address.

I'm proceeding with local testing meanwhile.

dplumlee commented on x-pack/solutions/security/plugins/security_solution/common/detection_engine/constants.ts:
```diff
@@ -74,3 +75,103 @@ export const defaultRiskScoreBySeverity: Record<Severity, number> = {
+
+type AllKeys<U> = U extends unknown ? keyof U : never;
+
+/**
+ * A list of all possible fields in the RuleResponse type mapped to whether or not the field is
+ * considered functional. We are defining "functional" to mean having a direct impact on how a
+ * rule executes. This means fields like `query` will be marked as functional while fields like
+ * `note` will be marked as non-functional. We are being conservative in our labeling of
+ * functional and only fields that have a 100% guaranteed impact on rule execution will be labeled
+ * as such. Fields like `index` that have a direct impact but don't necessarily change the alert
+ * rate (noise) of a rule will not be marked as functional.
+ *
+ * This categorization is intended to be used for telemetry purposes.
+ *
+ * More info here:
+ * x-pack/solutions/security/plugins/security_solution/docs/rfcs/detection_response/customized_rule_alert_telemetry.md
+ */
+export const FUNCTIONAL_FIELD_MAP: Record<AllKeys<RuleResponse>, boolean> = {
```
Moved over

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

dplumlee committed:
addresses comments

marshallmain commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/converters/common_params_camel_to_snake.ts:
Yeah we shouldn't keep both functions around long term. Since the only usage is now for alert building and is intended to store an exact snapshot of the rule parameters, just snake cased instead of camel cased, we may be able to replace that usage with the generic `convertObjectKeysToSnakeCase` instead - which would let us remove `commonParamsCamelToSnake` entirely. However, I generally oppose using `convertObjectKeysToSnakeCase` for the SO <-> API conversions because it lacks type safety.

marshallmain replied:
> Nice! So you're saying, besides the casing conversion, we wouldn't even need to normalize for the purpose of storing params in alerts? That makes it very simple and easy.

I think we'd want to match the normalization behavior that's currently in `commonParamsCamelToSnake`, but it's fairly simple normalization so hopefully less of a maintenance issue than maintaining the whole function in its current form. Otherwise, we could try to track down all the ways we use the rule params in alerts and make sure the normalization isn't actually required, but that seems harder than just keeping the existing normalization behavior. 🤷

banderror reviewed (approved):
@dplumlee This fixed implementation looks correct to me, thanks for the fixes and the additional test coverage 👍 

I left just a couple nits for your consideration.

I'm trying to test this locally, just a happy path. I can confirm that a customized rule gets a `rule_source` with all the fields correctly initialized. I can also see that telemetry is being sent. But I'm not seeing it in the `analytics-staging` cluster. How much time does it take to process and index the incoming telemetry events? Is there any sampling that requires sending a lot of it so that at least some gets indexed?

Approving the PR in advance because the thoroughly reviewed implementation looks correct and the test coverage seems robust enough. Thanks for the hard work @dplumlee!

dplumlee committed:
addresses comments

dplumlee committed:
Merge remote-tracking branch 'upstream/main' into rule-customizations-field

dplumlee committed:
fix message

elasticmachine committed:
Merge branch 'main' into rule-customizations-field

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/346668)
* Commit: 8eab26608bf176a89471ea2fbed64ad56465c3c0

### Failed CI Steps
* [FTR Configs #32](https://buildkite.com/elastic/kibana-pull-request/builds/346668#0199a4d1-601f-4833-905e-5ff8d74faac5)
* [FTR Configs #58](https://buildkite.com/elastic/kibana-pull-request/builds/346668#0199a4d1-6037-47cf-b932-fdb24fb9b09f)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/346668#0199a4d1-6037-47cf-b932-fdb24fb9b09f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/346668/jobs/0199a4d1-6037-47cf-b932-fdb24fb9b09f/artifacts/0199a4fe-b4f7-440a-8730-480496539d0a) FTR Configs #<span></span>58 / Space awareness agents POST /agents/bulk_upgrade should only upgrade agents in the same space when passing a kuery

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/064f1d5d38994113bb81008da708b5d1841c6586) | [after](https://github.com/elastic/kibana/commit/8eab26608bf176a89471ea2fbed64ad56465c3c0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 11.1MB | 11.1MB | +110.0B |



### History
* :green_heart: [Build #346471](https://buildkite.com/elastic/kibana-pull-request/builds/346471) succeeded 1caa2974aaa352fef894e357070751c108a177c7
* :broken_heart: [Build #346338](https://buildkite.com/elastic/kibana-pull-request/builds/346338) failed cb163a4c970756fe89c4c7c74ebcfe50ad703c2b
* :broken_heart: [Build #346049](https://buildkite.com/elastic/kibana-pull-request/builds/346049) failed 664aca0ad6a0cdda11ea697633f29f6ac27376c2
* :broken_heart: [Build #346004](https://buildkite.com/elastic/kibana-pull-request/builds/346004) failed 14d09a9d878f4ff190629ec7d161981630b69fe6
* :broken_heart: [Build #345844](https://buildkite.com/elastic/kibana-pull-request/builds/345844) failed 0f9c568b4790256cb2189e45849b4da6a474e4ef
* :broken_heart: [Build #345485](https://buildkite.com/elastic/kibana-pull-request/builds/345485) failed 555f4d7dccd8c8d70ed717fa052558722c36a533

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346668","number":346668,"commit":"8eab26608bf176a89471ea2fbed64ad56465c3c0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346471","number":346471,"commit":"1caa2974aaa352fef894e357070751c108a177c7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346338","number":346338,"commit":"cb163a4c970756fe89c4c7c74ebcfe50ad703c2b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346049","number":346049,"commit":"664aca0ad6a0cdda11ea697633f29f6ac27376c2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346004","number":346004,"commit":"14d09a9d878f4ff190629ec7d161981630b69fe6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345844","number":345844,"commit":"0f9c568b4790256cb2189e45849b4da6a474e4ef"}],"number":346668}
buildkite-pr-comment-->

dplumlee committed:
Merge branch 'main' into rule-customizations-field

dplumlee committed:
Merge branch 'main' into rule-customizations-field

kibanamachine commented:
Starting backport for target branches: 9.1, 9.2

https://github.com/elastic/kibana/actions/runs/18232333038

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/237536">](https://github.com/elastic/kibana/pull/237536)|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/237537">](https://github.com/elastic/kibana/pull/237537)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
